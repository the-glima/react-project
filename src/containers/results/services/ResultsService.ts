import {ResultsResponse, ResultsTier, ResultsWinningNumbers} from '../models/ResultsResponseModel'
import {ResultsDOM, ResultsDOMDate} from '../models/ResultsDOMModel'
import {settings} from '../../../settings'

const ResultsService = {
  fetchData: async function (url = settings.api.url): Promise<ResultsResponse> {
    try {
      const response = await fetch(url)
      return await response.json()
    } catch (error) {
      throw error
    }
  },

  sortTiers: function (tiers: ResultsTier): any[] {
    return Object.entries(tiers).sort((a, b) => a[0].localeCompare(b[0], 'en', {numeric: true}))
  },

  mapTiers: function (tiers: any): ResultsTier[] {
    const sortedTiers: ResultsTier[] = this.sortTiers(tiers)

    return Object.values(sortedTiers).map((tier: any) => ({
      winners: tier[1].winners,
      specialPrize: tier[1].specialPrize,
      prize: tier[1].prize
    }))
  },

  getResults: async function (): Promise<ResultsDOM> {
    const {last, next} = await this.fetchData()
    const tiers = last?.odds ? this.mapTiers(last.odds) : []

    return {
      currency: last.currency,
      last: {
        id: last.nr,
        date: {
          day: last.date.day,
          month: last.date.month,
          year: last.date.year,
          hour: last.date.hour,
          minute: last.date.minute
        },
        drawingDate: last.drawingDate,
        winningNumbers: {
          numbers: last.numbers,
          euroNumbers: last.euroNumbers
        },
        tiers: [...tiers]
      },
      next: {
        id: next.nr,
        date: {
          day: last.date.day,
          month: last.date.month,
          year: last.date.year,
          hour: last.date.hour,
          minute: last.date.minute
        },
        drawingDate: next.drawingDate
      }
    }
  },

  getResultWinners: function (winners: ResultsTier[] | undefined): any {
    const data = winners ? Object.values(winners) : null

    return data && data.length ? data.filter((item: any) => item.winners > 0) : null
  },

  getWinningNumbers: function (numbers: ResultsWinningNumbers | undefined): any {
    return numbers ? numbers : null
  },

  getResultDate: function (date: ResultsDOMDate | undefined, separator = '/'): any {
    if (!date) return null

    const addZero = (number: number) => (number && number < 10 ? `0${number}` : number)
    const dateFormatted = date && `${addZero(date.month)}${separator}${addZero(date.day)}${separator}${date.year}`

    return dateFormatted
  }
}

export default ResultsService
