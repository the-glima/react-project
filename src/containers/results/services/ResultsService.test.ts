import ResultsService from './ResultsService'
import {ResultsMock} from '../../../test/mocks/ResultsMock'
import fetchMock from 'jest-fetch-mock'
import {enableFetchMocks} from 'jest-fetch-mock'

describe('Service: Results', () => {
  const url = 'https://example.com'

  beforeEach(() => {
    enableFetchMocks()

    fetchMock.resetMocks()
  })

  it('fetchData: should fetch and return json data', async (done) => {
    fetchMock.mockResponseOnce(JSON.stringify(ResultsMock))

    const result = await ResultsService.fetchData(url)

    expect(result).toMatchSnapshot()
    expect(fetchMock).toHaveBeenCalledTimes(1)
    expect(fetchMock).toHaveBeenCalledWith(url)
    done()
  })

  it('fetchData: should thrown an error', async (done) => {
    const error = new Error('Something went wrong')
    fetchMock.mockRejectOnce(error)

    const result = await ResultsService.fetchData(url)

    expect(result).toBeInstanceOf(Error)
    expect(result.toString()).toEqual('Error: Something went wrong')
    expect(fetchMock).toHaveBeenCalledTimes(1)
    expect(fetchMock).toHaveBeenCalledWith(url)
    done()
  })

  it('sortTiers: should sort tiers in ascending order', () => {
    const tiers = ResultsMock.last.odds
    const result = ResultsService.sortTiers(tiers)

    expect(result).toHaveLength(Object.entries(tiers).length)
    expect(result).toMatchSnapshot()
  })

  it('mapTiers: should map tiers', () => {
    const tiers = ResultsMock.last.odds
    const result = ResultsService.mapTiers(tiers as any)

    expect(result).toHaveLength(Object.entries(tiers).length)
    expect(result).toMatchSnapshot()
  })

  it('getResults: should transform and return results', async (done) => {
    fetchMock.mockResponseOnce(JSON.stringify(ResultsMock))

    const result = await ResultsService.getResults()

    expect(result).toMatchSnapshot()
    done()
  })

  it('filterWinners: should filters winners removing 0 value item', () => {
    const tiers = [
      {
        winners: 0,
        specialPrize: 0,
        prize: 0
      },
      {
        winners: 1,
        specialPrize: 1,
        prize: 1
      },
      {
        winners: 2,
        specialPrize: 2,
        prize: 2
      }
    ]

    const resultNull = ResultsService.filterWinners(undefined)
    expect(resultNull).toBeNull()

    const result = ResultsService.filterWinners(tiers)
    expect(result).toHaveLength(2)
    expect(result[0].winners).toBe(1)
    expect(result[1].winners).toBe(2)
    expect(result).toMatchSnapshot()
  })

  it('getWinningNumbers: should return numbers or null', () => {
    const numbers = {
      numbers: [1, 2, 3, 4],
      euroNumbers: [5, 6, 7]
    }
    const resultNull = ResultsService.getWinningNumbers(undefined)
    expect(resultNull).toBeNull()

    const result = ResultsService.getWinningNumbers(numbers)
    expect(result).toMatchSnapshot()
  })
})
