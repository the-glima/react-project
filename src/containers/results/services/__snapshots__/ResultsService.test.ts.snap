// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service: Results fetchData: should fetch and return json data 1`] = `
Object {
  "last": Object {
    "Winners": 812567,
    "climbedSince": 9,
    "closingDate": "08.05.2020, 19:00",
    "currency": "EUR",
    "date": Object {
      "day": 8,
      "dayOfWeek": "Freitag",
      "full": "Die Lottozahlen vom Freitag, den 08.05.2020",
      "hour": 21,
      "minute": 0,
      "month": 5,
      "year": 2020,
    },
    "drawingDate": "08.05.2020, 21:00",
    "euroNumbers": Array [
      8,
      9,
    ],
    "jackpot": "10",
    "lateClosingDate": "08.05.2020, 20:10",
    "marketingJackpot": "10",
    "nr": 425,
    "numbers": Array [
      9,
      11,
      15,
      36,
      43,
    ],
    "odds": Object {
      "rank0": Object {
        "prize": 0,
        "specialPrize": 0,
        "winners": 0,
      },
      "rank1": Object {
        "prize": 1328694340,
        "specialPrize": 0,
        "winners": 1,
      },
      "rank10": Object {
        "prize": 1520,
        "specialPrize": 0,
        "winners": 62773,
      },
      "rank11": Object {
        "prize": 1280,
        "specialPrize": 0,
        "winners": 135108,
      },
      "rank12": Object {
        "prize": 770,
        "specialPrize": 0,
        "winners": 546864,
      },
      "rank2": Object {
        "prize": 47210980,
        "specialPrize": 0,
        "winners": 4,
      },
      "rank3": Object {
        "prize": 8331350,
        "specialPrize": 0,
        "winners": 8,
      },
      "rank4": Object {
        "prize": 482970,
        "specialPrize": 0,
        "winners": 46,
      },
      "rank5": Object {
        "prize": 23770,
        "specialPrize": 0,
        "winners": 841,
      },
      "rank6": Object {
        "prize": 10760,
        "specialPrize": 0,
        "winners": 1444,
      },
      "rank7": Object {
        "prize": 7500,
        "specialPrize": 0,
        "winners": 1777,
      },
      "rank8": Object {
        "prize": 2670,
        "specialPrize": 0,
        "winners": 25727,
      },
      "rank9": Object {
        "prize": 1750,
        "specialPrize": 0,
        "winners": 37974,
      },
    },
    "specialMarketingJackpot": "10",
  },
  "next": Object {
    "climbedSince": 9,
    "closingDate": "15.05.2020, 19:00",
    "currency": "EUR",
    "date": Object {
      "day": 15,
      "dayOfWeek": "Freitag",
      "full": "Die Lottozahlen vom Freitag, den 15.05.2020",
      "hour": 21,
      "minute": 0,
      "month": 5,
      "year": 2020,
    },
    "drawingDate": "15.05.2020, 21:00",
    "jackpot": "10",
    "lateClosingDate": "15.05.2020, 20:10",
    "marketingJackpot": "10",
    "nr": 426,
    "specialMarketingJackpot": "10",
  },
}
`;

exports[`Service: Results filterWinners: should filters winners removing 0 value item 1`] = `
Array [
  Object {
    "prize": 1328694340,
    "specialPrize": 0,
    "winners": 1,
  },
  Object {
    "prize": 47210980,
    "specialPrize": 0,
    "winners": 4,
  },
  Object {
    "prize": 8331350,
    "specialPrize": 0,
    "winners": 8,
  },
  Object {
    "prize": 482970,
    "specialPrize": 0,
    "winners": 46,
  },
  Object {
    "prize": 23770,
    "specialPrize": 0,
    "winners": 841,
  },
  Object {
    "prize": 10760,
    "specialPrize": 0,
    "winners": 1444,
  },
  Object {
    "prize": 7500,
    "specialPrize": 0,
    "winners": 1777,
  },
  Object {
    "prize": 2670,
    "specialPrize": 0,
    "winners": 25727,
  },
  Object {
    "prize": 1750,
    "specialPrize": 0,
    "winners": 37974,
  },
  Object {
    "prize": 1520,
    "specialPrize": 0,
    "winners": 62773,
  },
  Object {
    "prize": 1280,
    "specialPrize": 0,
    "winners": 135108,
  },
  Object {
    "prize": 770,
    "specialPrize": 0,
    "winners": 546864,
  },
]
`;

exports[`Service: Results getResults: should transform and return results 1`] = `
Object {
  "currency": "EUR",
  "last": Object {
    "date": Object {
      "day": 8,
      "hour": 21,
      "minute": 0,
      "month": 5,
      "year": 2020,
    },
    "drawingDate": "08.05.2020, 21:00",
    "id": 425,
    "tiers": Array [
      Object {
        "prize": 0,
        "specialPrize": 0,
        "winners": 0,
      },
      Object {
        "prize": 1328694340,
        "specialPrize": 0,
        "winners": 1,
      },
      Object {
        "prize": 47210980,
        "specialPrize": 0,
        "winners": 4,
      },
      Object {
        "prize": 8331350,
        "specialPrize": 0,
        "winners": 8,
      },
      Object {
        "prize": 482970,
        "specialPrize": 0,
        "winners": 46,
      },
      Object {
        "prize": 23770,
        "specialPrize": 0,
        "winners": 841,
      },
      Object {
        "prize": 10760,
        "specialPrize": 0,
        "winners": 1444,
      },
      Object {
        "prize": 7500,
        "specialPrize": 0,
        "winners": 1777,
      },
      Object {
        "prize": 2670,
        "specialPrize": 0,
        "winners": 25727,
      },
      Object {
        "prize": 1750,
        "specialPrize": 0,
        "winners": 37974,
      },
      Object {
        "prize": 1520,
        "specialPrize": 0,
        "winners": 62773,
      },
      Object {
        "prize": 1280,
        "specialPrize": 0,
        "winners": 135108,
      },
      Object {
        "prize": 770,
        "specialPrize": 0,
        "winners": 546864,
      },
    ],
    "winningNumbers": Object {
      "euroNumbers": Array [
        8,
        9,
      ],
      "numbers": Array [
        9,
        11,
        15,
        36,
        43,
      ],
    },
  },
  "next": Object {
    "date": Object {
      "day": 8,
      "hour": 21,
      "minute": 0,
      "month": 5,
      "year": 2020,
    },
    "drawingDate": "15.05.2020, 21:00",
    "id": 426,
  },
}
`;

exports[`Service: Results getWinningNumbers: should return numbers or null 1`] = `
Object {
  "euroNumbers": Array [
    5,
    6,
    7,
  ],
  "numbers": Array [
    1,
    2,
    3,
    4,
  ],
}
`;

exports[`Service: Results mapTiers: should map tiers 1`] = `
Array [
  Object {
    "prize": 0,
    "specialPrize": 0,
    "winners": 0,
  },
  Object {
    "prize": 1328694340,
    "specialPrize": 0,
    "winners": 1,
  },
  Object {
    "prize": 47210980,
    "specialPrize": 0,
    "winners": 4,
  },
  Object {
    "prize": 8331350,
    "specialPrize": 0,
    "winners": 8,
  },
  Object {
    "prize": 482970,
    "specialPrize": 0,
    "winners": 46,
  },
  Object {
    "prize": 23770,
    "specialPrize": 0,
    "winners": 841,
  },
  Object {
    "prize": 10760,
    "specialPrize": 0,
    "winners": 1444,
  },
  Object {
    "prize": 7500,
    "specialPrize": 0,
    "winners": 1777,
  },
  Object {
    "prize": 2670,
    "specialPrize": 0,
    "winners": 25727,
  },
  Object {
    "prize": 1750,
    "specialPrize": 0,
    "winners": 37974,
  },
  Object {
    "prize": 1520,
    "specialPrize": 0,
    "winners": 62773,
  },
  Object {
    "prize": 1280,
    "specialPrize": 0,
    "winners": 135108,
  },
  Object {
    "prize": 770,
    "specialPrize": 0,
    "winners": 546864,
  },
]
`;

exports[`Service: Results sortTiers: should sort tiers in ascending order 1`] = `
Array [
  Array [
    "rank0",
    Object {
      "prize": 0,
      "specialPrize": 0,
      "winners": 0,
    },
  ],
  Array [
    "rank1",
    Object {
      "prize": 1328694340,
      "specialPrize": 0,
      "winners": 1,
    },
  ],
  Array [
    "rank2",
    Object {
      "prize": 47210980,
      "specialPrize": 0,
      "winners": 4,
    },
  ],
  Array [
    "rank3",
    Object {
      "prize": 8331350,
      "specialPrize": 0,
      "winners": 8,
    },
  ],
  Array [
    "rank4",
    Object {
      "prize": 482970,
      "specialPrize": 0,
      "winners": 46,
    },
  ],
  Array [
    "rank5",
    Object {
      "prize": 23770,
      "specialPrize": 0,
      "winners": 841,
    },
  ],
  Array [
    "rank6",
    Object {
      "prize": 10760,
      "specialPrize": 0,
      "winners": 1444,
    },
  ],
  Array [
    "rank7",
    Object {
      "prize": 7500,
      "specialPrize": 0,
      "winners": 1777,
    },
  ],
  Array [
    "rank8",
    Object {
      "prize": 2670,
      "specialPrize": 0,
      "winners": 25727,
    },
  ],
  Array [
    "rank9",
    Object {
      "prize": 1750,
      "specialPrize": 0,
      "winners": 37974,
    },
  ],
  Array [
    "rank10",
    Object {
      "prize": 1520,
      "specialPrize": 0,
      "winners": 62773,
    },
  ],
  Array [
    "rank11",
    Object {
      "prize": 1280,
      "specialPrize": 0,
      "winners": 135108,
    },
  ],
  Array [
    "rank12",
    Object {
      "prize": 770,
      "specialPrize": 0,
      "winners": 546864,
    },
  ],
]
`;
